benchmarks/linalg_refactor_benchmark.cpp

```diff
@@ -0,0 +1,151 @@
+#include <shogun/base/init.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/linalgrefactor/linalgRefactor.h>
+
+#include <shogun/mathematics/eigen3.h>
+#include <viennacl/linalg/inner_prod.hpp>
+#include <viennacl/vector.hpp>
+
+#include <algorithm>
+#include <memory>
+#include <hayai/hayai.hpp>
+
+#include <iostream>
+
+using namespace shogun;
+
+/**
+ * Instructions :
+ * 1. Install benchmarking toolkit "hayai" (https://github.com/nickbruun/hayai)
+ * 2. Compile against libhayai_main, e.g.
+ * 		g++ -O3 -std=c++11 linglg_refactor_benchmark.cpp -I/usr/include/eigen3 \
+ *		-lshogun -lhayai_main -lOpenCL -o benchmark
+ * 3. ./benchmark
+ */
+
+/** Generate data only once */
+typedef float32_t T;
+typedef viennacl::vector_base<T, std::size_t, std::ptrdiff_t> VCLVectorBase;
+
+template<typename value_type>
+struct Data
+{
+	typedef viennacl::backend::mem_handle VCLMemoryArray;
+
+	Data()
+	{
+		num_rows = 100;
+		init();
+	}
+
+	Data(index_t num_rows)
+	{
+		this->num_rows = num_rows;
+		init();
+	}
+
+	~Data()
+	{
+		exit_shogun();
+	}
+
+	void init()
+	{
+		A = init_sg(1);
+		B = init_sg(2);
+		Ac = init_c(A);
+		Bc = init_c(B);
+		Ag = init_g(A);
+		Bg = init_g(B);
+		Av = init_v(A);
+		Bv = init_v(B);
+		init_shogun_with_defaults();
+		GPUBackend viennaclBackend;
+		sg_linalg->set_gpu_backend(&viennaclBackend);
+	}
+
+	/** SGVector **/
+	SGVector<value_type> init_sg(value_type begin)
+	{
+		SGVector<value_type> m(num_rows);
+		m.range_fill(begin);
+
+		return m;
+	}
+
+	/** CPUVector derived from BaseVector **/
+	CPUVector<value_type> init_c(SGVector<value_type> m)
+	{
+        	CPUVector<value_type> mc(m);
+        	return mc;
+	}
+
+	/** GPUVector derived from BaseVector **/
+	GPU_Vector<value_type> init_g(SGVector<value_type> m)
+	{
+        	GPU_Vector<value_type> mg(m);
+        	return mg;
+	}
+
+	/** ViennaCL Vector for test **/
+	VCLVectorBase init_v(SGVector<value_type> m)
+	{
+		VCLVectorBase mv;
+		std::shared_ptr<VCLMemoryArray> vector(new VCLMemoryArray());
+		viennacl::backend::memory_create(*vector, sizeof(value_type)*num_rows,
+			viennacl::context());
+		viennacl::backend::memory_write(*vector, 0, num_rows*sizeof(value_type),
+			m.vector);
+		mv = VCLVectorBase(*vector, num_rows, 0, 1);
+
+		return mv;
+	}
+
+	SGVector<value_type> A;
+	SGVector<value_type> B;
+	CPUVector<value_type> Ac;
```

Comment:
Just have a `BaseVector<value_type>* Ac` here. Same for others. Should be allocated in the heap.